internal/pkg/skuba/replica/replica.go

```diff
@@ -0,0 +1,409 @@
+func (r *Helper) updateDeploymentReplica(size int) (*appsv1.Deployment, error) {
+	replicaJSON := fmt.Sprintf(patchReplicas, size)
+	return r.Client.AppsV1().Deployments(metav1.NamespaceSystem).Patch(r.Deployment.ObjectMeta.Name, types.StrategicMergePatchType, []byte(replicaJSON))
+}
+
+// UpdateNodes patches for replicas affinity rules after adding of nodes.
+// This updates affinity rule to preferredDuringSchedulingIgnoredDuringExecution of
+// cluster size less then replica size,
+// And updates affinity rule to requiredDuringSchedulingIgnoredDuringExecution when
+// cluster size is equal or greater to replicas.
+func (r *Helper) UpdateNodes() error {
+	if err := r.deploymentsHelper(); err != nil {
+		return err
+	}
+
+	for _, deployment := range r.Deployments.Items {
+		if err := r.deploymentHelper(deployment.Name); err != nil {
+			return err
+		}
+		r.ReplicaSize = int(*r.Deployment.Spec.Replicas)
+		if r.ReplicaSize == 0 {
+			r.ReplicaSize = r.MinSize
+		}
+
+		nodes := r.ClusterSize
+		updated := false
+		var err error
+		switch {
+		case nodes >= r.ReplicaSize:
+			updated, err = r.replaceAffinity(preferred, required)
+			if err != nil {
+				return err
+			}
+		case nodes >= r.MinSize:
+			updated, err = r.replaceAffinity(required, preferred)
+			if err != nil {
+				return err
+			}
+		}
+		if !updated && nodes <= r.ReplicaSize {
+			// update replicas to trigger pod re-distribution when affinity is in
+			// preferredDuringSchedulingIgnoredDuringExecution.
+			if _, err := r.updateDeploymentReplica(r.ClusterSize - 1); err != nil {
+				return err
+			}
+			if _, err := r.updateDeploymentReplica(r.ReplicaSize); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// UpdateBeforeNodeDrains patches affinity rules and wait for pod running before node remove.
+// Patch will apply to deployment with preferredDuringSchedulingIgnoredDuringExecution affinity
+// when cluster size equal or less then replicas.
+func (r *Helper) UpdateBeforeNodeDrains() error {
+	if err := r.deploymentsHelper(); err != nil {
+		return err
+	}
+
+	for _, deployment := range r.Deployments.Items {
+		if err := r.deploymentHelper(deployment.Name); err != nil {
+			return err
+		}
+		r.ReplicaSize = int(*r.Deployment.Spec.Replicas)
+		if r.ClusterSize > r.ReplicaSize {
+			continue
+		}
+
+		updated, err := r.replaceAffinity(required, preferred)
+		if err != nil {
+			return err
+		}
+		if !updated && r.ClusterSize <= r.ReplicaSize {
+			// update replicas to trigger pod re-distribution when affinity is in
+			// preferredDuringSchedulingIgnoredDuringExecution.
+			if _, err := r.updateDeploymentReplica(r.ClusterSize - 1); err != nil {
+				return err
+			}
+			if _, err := r.updateDeploymentReplica(r.ReplicaSize); err != nil {
+				return err
+			}
+			updated = true
+		}
+		if updated {
+			var err error
+			retry := retryCount
+			err = r.waitForUpdates(&retry, &err)
+			if err != nil {
+				return err
+			}
```

Comment:
Please change to below. 

```
if err := r.waitForUpdates(retryCount); err != nil {
    return err
}
```

Reply:
I am thinking if we should swallow this error w/ a warning message instead of throwing out to cause operation failed.

Because the change already accepted by API server, just need to remind user pod anti-affinity still taking progress asynchronously. But for test cases, it should be enhanced to add some checking login between the steps.

What do you think?

Reply:
IMO I don't really agree to swallowing the error while reboot is problematic. The wait error most likely to cause by exhaust retried due to slow network connection while doing image pull during reallocation.
The operation is idempotent, so user can  run the same remove command again to ensure pods are allocated correctly. If it always return fail, there might be some other underlying issues which need manual fix.

Reply:
Let me correct it’s not swallowing the error but instead process by logging warning instead of bubble up.

But as we discussed, we need to make sure idempotent operation. I would suggest adding more test cases for node reboot/shutdown in caasp-test-cases repo. Thanks!

Reply:
> Please change to below.
> 
> ```
> if err := r.waitForUpdates(retryCount); err != nil {
>     return err
> }
> ```

done

Reply:
> Let me correct it’s not swallowing the error but instead process by logging warning instead of bubble up.
> 
> But as we discussed, we need to make sure idempotent operation. I would suggest adding more test cases for node reboot/shutdown in caasp-test-cases repo. Thanks!

[Here](https://github.com/maximenoel8/caasp_robotframework/blob/master/test/replica.robot) is robotframework test case by @maximenoel8.